src/DDTrace/OpenTelemetry/Span.php

```diff
@@ -0,0 +1,392 @@
+    {
+        // no-op
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function recordException(Throwable $exception, iterable $attributes = []): SpanInterface
+    {
+        if (!$this->hasEnded()) {
+            $this->span->meta[Tag::ERROR_MSG] = $exception->getMessage();
+            $this->span->meta[Tag::ERROR_TYPE] = get_class($exception);
+            $this->span->meta[Tag::ERROR_STACK] = $exception->getTraceAsString();
+        }
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function updateName(string $name): SpanInterface
+    {
+        // OTel.name => DD.resource
+        if (!$this->hasEnded()) {
+            $this->span->resource = $name;
+        }
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function setStatus(string $code, string $description = null): SpanInterface
+    {
+        if ($this->hasEnded()) {
+            return $this;
+        }
+
+        // An attempt to set value Unset SHOULD be ignored.
+        if ($code === API\StatusCode::STATUS_UNSET) {
+            return $this;
+        }
+
+        // When span status is set to Ok it SHOULD be considered final and any further attempts to change it SHOULD be ignored.
+        if ($this->status->getCode() === API\StatusCode::STATUS_OK) {
+            return $this;
+        }
+
+        if ($this->status->getCode() === API\StatusCode::STATUS_UNSET && $code === API\StatusCode::STATUS_ERROR) {
+            $this->span->meta[Tag::ERROR_MSG] = $description;
+        } elseif ($this->status->getCode() === API\StatusCode::STATUS_ERROR && $code === API\StatusCode::STATUS_OK) {
+            unset($this->span->meta[Tag::ERROR_MSG]);
+            unset($this->span->meta[Tag::ERROR_TYPE]);
+            unset($this->span->meta[Tag::ERROR_STACK]);
+        }
+
+        $this->status = StatusData::create($code, $description);
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function end(int $endEpochNanos = null): void
+    {
+        if ($this->hasEnded()) {
+            return;
+        }
+
+        $this->endOTelSpan($endEpochNanos);
+
+        switch_stack($this->span);
+        close_span($endEpochNanos !== null ? $endEpochNanos / 1000000000 : 0);
+    }
+
+    public function endOTelSpan(int $endEpochNanos = null): void
+    {
+        if ($this->hasEnded()) {
+            return;
+        }
+
+        if (empty($this->span->name)) { // Honor set operation name
+            $this->span->name = Convention::defaultOperationName($this->span);
+        }
+
+        // After closing the span, the context won't change, but would otherwise be lost
+        $this->context = API\SpanContext::create(
+            $this->context->getTraceId(),
+            $this->context->getSpanId(),
+            $this->context->getTraceFlags(),
+            $this->context->getTraceState()
+        );
+
+        $this->spanProcessor->onEnd($this);
```

Comment:
Not sure about this:
- Given that endOTelSpan is only called on Context::current(), if there's no further access to Context::current() after terminating the span, the spanProcessor->onEnd() will not be called
- this only runs for datadog-only Spans which happen to be accessed at least once by a Context::current(), right?
- I'm not sure what the best behaviour is, maybe, for consistency _exclusively_ run it on OTel managed spans?

Reply:
> Given that endOTelSpan is only called on Context::current(), if there's no further access to Context::current() after terminating the span, the spanProcessor->onEnd() will not be called

Yes. One alternative I also tried would have been to do an OpenTelemetryIntegration, which traces a span's termination... but faced limitations. To this day, I still think this would have been the cleanest way overall (i.e., have sort of an event when a DD span is created/closed without the OTel API so that we could simply instrument them and start/end OTel spans as needed).

> this only runs for datadog-only Spans, which happen to be accessed at least once by a Context::current(), right?

Yes

> I'm not sure what the best behaviour is, maybe, for consistency exclusively run it on OTel managed spans?

For consistency, it may indeed make sense